README.md
# holiday-calculator

[![npm-version](https://img.shields.io/npm/v/holiday-calculator.svg)](https://www.npmjs.com/package/holiday-calculator)
[![Build Status](https://travis-ci.org/darrinholst/holiday-calculator.svg?branch=master)](https://travis-ci.org/darrinholst/holiday-calculator)
[![dependencies](https://david-dm.org/darrinholst/holiday-calculator/status.svg)](https://david-dm.org/darrinholst/holiday-calculator)
[![dev dependencies](https://david-dm.org/darrinholst/holiday-calculator/dev-status.svg)](https://david-dm.org/darrinholst/holiday-calculator?type=dev)
[![peer dependencies](https://david-dm.org/darrinholst/holiday-calculator/peer-status.svg)](https://david-dm.org/darrinholst/holiday-calculator?type=peer)

Holiday utilities for Moment. Not a plugin though. 😒

### Installation

#### node

    npm install holiday-calculator

### API

##### `countBusinessHolidaysBetween(startDate: Moment | string, endDate: Moment | string)`

Returns the number of holidays observed by businesses between `startDate` and `endDate`.

``` js
import {US} from 'holiday-calculator';

new US().countBusinessHolidaysBetween('2017-01-01', '2017-12-31'); // => 6
```

### Holiday Rules

All the rules for determining holidays is encapsulated in a `Holidays` object. `US` is currently the
only one that's builtin. It recognizes:

1. New Years Day
1. Martin Luther King Jr. Day
1. Presidents' Day
1. Memorial Day
1. Independence Day
1. Labor Day
1. Veterans Day
1. Thanksgiving
1. Christmas Day

### Development
    git clone https://github.com/darrinholst/moment-holiday
    cd moment-holiday
    npm i
    npm test

### Release
    release-it [semver]


CHANGELOG.md
### [0.0.1-beta.4](https://github.com/darrinholst/holiday-calculator/releases/tag/0.0.1-beta.1)

* Added MLK day, Presidents Day & Veterans Day to US list

### [0.0.1-beta.2](https://github.com/darrinholst/holiday-calculator/releases/tag/0.0.1-beta.1)

* API changes to better reflect what the current holiday counter is doing.
* Require explicit use of a holiday config. i.e. `new US()`

### [0.0.1-beta.1](https://github.com/darrinholst/holiday-calculator/releases/tag/0.0.1-beta.1)

Changed release-it to a dev dependency.

### [0.0.1-beta.0](https://github.com/darrinholst/holiday-calculator/releases/tag/0.0.1-beta.0)

First beta release.


src/us.spec.ts
import * as moment from 'moment';
import 'mocha';
import {expect} from 'chai';

import {US} from './index';

describe('US Holidays', () => {
  let us = new US();

  describe('business days', () => {
    it('should count all the holidays for the year', () => {
      expect(
        us.countBusinessHolidaysBetween('2017-01-01', '2017-12-31')
      ).to.eql(9);
    });

    it('should count new years', function() {
      expect(
        us.countBusinessHolidaysBetween('2018-01-01', '2018-01-05')
      ).to.eql(1);
    });

    it('should count the monday after new years when it is on a sunday', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-01-02', '2017-01-06')
      ).to.eql(1);
    });

    it('should count the friday before new years when it is on a saturday', function() {
      expect(
        us.countBusinessHolidaysBetween('2021-12-27', '2021-12-31')
      ).to.eql(1);
    });

    it('should count independence day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-07-03', '2017-07-07')
      ).to.eql(1);
    });

    it('should count the monday after independence day when it is on a sunday', function() {
      expect(
        us.countBusinessHolidaysBetween('2015-06-29', '2015-07-03')
      ).to.eql(1);
    });

    it('should count the friday before independence day when it is on a saturday', function() {
      expect(
        us.countBusinessHolidaysBetween('2021-07-05', '2021-07-09')
      ).to.eql(1);
    });

    it('should count memorial day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-05-29', '2017-06-02')
      ).to.eql(1);
    });

    it('should count labor day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-09-04', '2017-09-08')
      ).to.eql(1);
    });

    it('should count thanksgiving', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-11-20', '2017-11-24')
      ).to.eql(1);
    });

    it('should count christmas', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-12-25', '2017-12-29')
      ).to.eql(1);
    });

    it('should count the monday after christmas when it is on a sunday', function() {
      expect(
        us.countBusinessHolidaysBetween('2010-12-20', '2010-12-24')
      ).to.eql(1);
    });

    it('should count the friday before christmas when it is on a saturday', function() {
      expect(
        us.countBusinessHolidaysBetween('2022-12-26', '2022-12-30')
      ).to.eql(1);
    });

    it('should count veterans day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-11-09', '2017-11-10')
      ).to.eql(1);
    });

    it('should count mlk day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-01-16', '2017-01-18')
      ).to.eql(1);
    });

    it('should count presidents day', function() {
      expect(
        us.countBusinessHolidaysBetween('2017-02-20', '2017-02-20')
      ).to.eql(1);
    });
  });
});


src/holidays.spec.ts
import * as moment from 'moment';
import 'mocha';
import {expect} from 'chai';

import {holidays} from './index';

describe('holidays', () => {
  it('should not find any holidays if nothing has been configured', () => {
    expect(holidays(moment('2017-01-01'), moment('2017-12-31'))).to.eql(0);
  });
});


src/sets/us-minimal.spec.ts
import * as moment from 'moment';
import 'mocha';
import {expect} from 'chai';

import {holidays, HolidaySet} from '../index';

describe('US Minimal Holidays', () => {
  it('should find all the holidays for the year', () => {
    expect(holidays(HolidaySet.UsMinimal, '2017-01-01', '2017-12-31')).to.eql(6);
  });

  it('should count new years', function() {
    expect(holidays(HolidaySet.UsMinimal, '2018-01-01', '2018-01-05')).to.eql(1);
  });

  it('should count the monday after new years when it is on a sunday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-01-02', '2017-01-06')).to.eql(1);
  });

  it('should count the friday before new years when it is on a saturday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2021-12-27', '2021-12-31')).to.eql(1);
  });

  it('should count independence day', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-07-03', '2017-07-07')).to.eql(1);
  });

  it('should count the monday after independence day when it is on a sunday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2015-06-29', '2015-07-03')).to.eql(1);
  });

  it('should count the friday before independence day when it is on a saturday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2021-07-05', '2021-07-09')).to.eql(1);
  });

  it('should count memorial day', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-05-29', '2017-06-02')).to.eql(1);
  });

  it('should count labor day', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-09-04', '2017-09-08')).to.eql(1);
  });

  it('should count thanksgiving', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-11-20', '2017-11-24')).to.eql(1);
  });

  it('should count christmas', function() {
    expect(holidays(HolidaySet.UsMinimal, '2017-12-25', '2017-12-29')).to.eql(1);
  });

  it('should count the monday after christmas when it is on a sunday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2010-12-20', '2010-12-24')).to.eql(1);
  });

  it('should count the friday before christmas when it is on a saturday', function() {
    expect(holidays(HolidaySet.UsMinimal, '2022-12-26', '2022-12-30')).to.eql(1);
  });
});


src/holidays.ts
import * as moment from 'moment';

export abstract class Holidays {
  abstract countBusinessHolidaysBetween(
    fromDate: moment.Moment | string,
    toDate: moment.Moment | string
  ): number;

  protected observedHoliday(date: string) {
    let momentized = moment(date);
    if (momentized.day() === 0) return momentized.add(1, 'day');
    if (momentized.day() === 6) return momentized.subtract(1, 'day');
    return momentized;
  }

  protected third(year: number, month: string, day: string) {
    return this.first(year, month, day).add(2, 'weeks');
  }

  protected fourth(year: number, month: string, day: string) {
    return this.first(year, month, day).add(3, 'weeks');
  }

  protected last(year: number, month: string, day: string) {
    return moment(`${year}-${month}-01`)
      .endOf('month')
      .startOf('isoWeek');
  }

  protected first(year: number, month: string, day: string) {
    let theFirstDayOfTheMonth = moment(`${year}-${month}-01`)
      .startOf('month')
      .day(day);
    if (theFirstDayOfTheMonth.date() > 7) theFirstDayOfTheMonth.add(7, 'days');
    return theFirstDayOfTheMonth;
  }
}


src/sets/us-minimal.ts
import * as moment from 'moment';

import {HolidaySet} from './holiday-set';

export class UsMinimal extends HolidaySet {
  private yearsInitialized: {[index: number]: boolean} = {};
  private observedHolidays: moment.Moment[] = [];

  count(fromDate: moment.Moment, toDate: moment.Moment) {
    this.initialize(fromDate.year() - 1, toDate.year() + 1);

    return this.observedHolidays.reduce((count, holiday) => {
      return count + (holiday.isBetween(fromDate, toDate, 'day', '[]') ? 1 : 0);
    }, 0)
  }

  private newYears = (year: number) => this.weekendableHoliday(`${year}-01-01`);
  private memorialDay = (year: number) => moment(`${year}-05-01`).endOf('month').startOf('isoWeek');
  private independenceDay = (year: number) => this.weekendableHoliday(`${year}-07-04`);
  private laborDay = (year: number) => this.first(year, '09', 'Monday');
  private thanksgiving = (year: number) => this.first(year, '11', 'Thursday').add(3, 'weeks');
  private christmas = (year: number) => this.weekendableHoliday(`${year}-12-25`);

  private initialize(fromYear: number, toYear: number) {
    for(let year = fromYear; year <= toYear; year++) this.initializeHolidaysForYear(year);
  }

  private initializeHolidaysForYear(year: number) {
    if (this.yearsInitialized[year]) return;
    this.yearsInitialized[year] = true;
    this.observedHolidays.push(this.newYears(year));
    this.observedHolidays.push(this.memorialDay(year));
    this.observedHolidays.push(this.independenceDay(year));
    this.observedHolidays.push(this.laborDay(year));
    this.observedHolidays.push(this.thanksgiving(year));
    this.observedHolidays.push(this.christmas(year));
  }
}


src/us.ts
import * as moment from 'moment';

import {Holidays} from './holidays';

export class US extends Holidays {
  countBusinessHolidaysBetween(
    fromDate: moment.Moment | string,
    toDate: moment.Moment | string
  ) {
    fromDate = moment(fromDate);
    toDate = moment(toDate);
    this.initialize(fromDate.year() - 1, toDate.year() + 1);

    return this.observedHolidays.reduce((count, holiday) => {
      return count + (holiday.isBetween(fromDate, toDate, 'day', '[]') ? 1 : 0);
    }, 0);
  }

  private yearsInitialized: {[index: number]: boolean} = {};
  private observedHolidays: moment.Moment[] = [];
  private newYears = (year: number) => this.observedHoliday(`${year}-01-01`);
  private mlk = (year: number) => this.third(year, '01', 'Monday');
  private presidents = (year: number) => this.third(year, '02', 'Monday');
  private memorialDay = (year: number) => this.last(year, '05', 'Monday');
  private july4th = (year: number) => this.observedHoliday(`${year}-07-04`);
  private laborDay = (year: number) => this.first(year, '09', 'Monday');
  private veterans = (year: number) => this.observedHoliday(`${year}-11-11`);
  private thanksgiving = (year: number) => this.fourth(year, '11', 'Thursday');
  private christmas = (year: number) => this.observedHoliday(`${year}-12-25`);

  private initialize(fromYear: number, toYear: number) {
    for (let year = fromYear; year <= toYear; year++) {
      this.initializeHolidaysForYear(year);
    }
  }

  private initializeHolidaysForYear(year: number) {
    if (this.yearsInitialized[year]) return;
    this.yearsInitialized[year] = true;
    this.observedHolidays.push(this.newYears(year));
    this.observedHolidays.push(this.mlk(year));
    this.observedHolidays.push(this.presidents(year));
    this.observedHolidays.push(this.memorialDay(year));
    this.observedHolidays.push(this.july4th(year));
    this.observedHolidays.push(this.laborDay(year));
    this.observedHolidays.push(this.veterans(year));
    this.observedHolidays.push(this.thanksgiving(year));
    this.observedHolidays.push(this.christmas(year));
  }
}
